nixos/modules/installer/cd-dvd/iso-image.nix

```diff
@@ -76,49 +156,136 @@ let
+    '' else ''
+      if background_image (hd0)/EFI/boot/efi-background.png; then
+        # Black background means transparent background when there
+        # is a background image set... This seems undocumented :(
+        set color_normal=black/black
+        set color_highlight=white/blue
+      else
+        # Falls back again to proper colors.
+        set menu_color_normal=cyan/blue
+        set menu_color_highlight=white/blue
+      fi
+    ''}
+  '';
+
   # The EFI boot image.
+  # Notes about grub:
+  #  * Yes, the grubMenuCfg has to be repeated in all submenus.
   efiDir = pkgs.runCommand "efi-directory" {} ''
-    mkdir -p $out/EFI/boot
-    cp -v ${pkgs.systemd}/lib/systemd/boot/efi/systemd-boot${targetArch}.efi $out/EFI/boot/boot${targetArch}.efi
-    mkdir -p $out/loader/entries
-
-    cat << EOF > $out/loader/entries/nixos-iso.conf
-    title NixOS ${config.system.nixos.label}${config.isoImage.appendToMenuLabel}
-    linux /boot/${config.system.boot.loader.kernelFile}
-    initrd /boot/${config.system.boot.loader.initrdFile}
-    options init=${config.system.build.toplevel}/init ${toString config.boot.kernelParams}
-    EOF
-
-    # A variant to boot with 'nomodeset'
-    cat << EOF > $out/loader/entries/nixos-iso-nomodeset.conf
-    title NixOS ${config.system.nixos.label}${config.isoImage.appendToMenuLabel}
-    version nomodeset
-    linux /boot/${config.system.boot.loader.kernelFile}
-    initrd /boot/${config.system.boot.loader.initrdFile}
-    options init=${config.system.build.toplevel}/init ${toString config.boot.kernelParams} nomodeset
-    EOF
-
-    # A variant to boot with 'copytoram'
-    cat << EOF > $out/loader/entries/nixos-iso-copytoram.conf
-    title NixOS ${config.system.nixos.label}${config.isoImage.appendToMenuLabel}
-    version copytoram
-    linux /boot/${config.system.boot.loader.kernelFile}
-    initrd /boot/${config.system.boot.loader.initrdFile}
-    options init=${config.system.build.toplevel}/init ${toString config.boot.kernelParams} copytoram
-    EOF
-
-    # A variant to boot with verbose logging to the console
-    cat << EOF > $out/loader/entries/nixos-iso-debug.conf
-    title NixOS ${config.system.nixos.label}${config.isoImage.appendToMenuLabel} (debug)
-    linux /boot/${config.system.boot.loader.kernelFile}
-    initrd /boot/${config.system.boot.loader.initrdFile}
-    options init=${config.system.build.toplevel}/init ${toString config.boot.kernelParams} loglevel=7
+    mkdir -p $out/EFI/boot/
+    
+    MODULES="fat iso9660 part_gpt part_msdos \
+             normal boot linux configfile loopback chain halt \
+             efifwsetup efi_gop efi_uga \
+             ls search search_label search_fs_uuid search_fs_file \
+             gfxmenu gfxterm gfxterm_background gfxterm_menu test all_video loadenv \
+             exfat ext2 ntfs btrfs hfsplus udf \
+             videoinfo png \
+            "
+    # Make our own efi program, we can't rely on `grub-install` since it seems to
```

Comment:
This wouldn't have been an issue with `installer/cd-dvd/iso-image.nix` before since `systemd-boot` was used, and not grub.

As for elsewhere, there is no use of the `--skip-fs-probe` parameter, and all other uses of `grub-install`, I don't think they're used in the same use case; building a disk image in a sandbox. I remember having to patch grub to find out which instance of the error message (failed to get canonical path of ...) was the one printed in the terminal, and finding out it was from the `grub-probe.c / probe()` function. That is because the path where grub is installed is not on a `/dev/disk/*` device, as far as I could tell.

In the end, it turns out `grub-install` may be an end-user tool, while integrators would prefer working directly with the other tools.

Reply:
> This wouldn't have been an issue with installer/cd-dvd/iso-image.nix before since systemd-boot was used, and not grub.

Oh I see, so it does change the default for the installer, but not NixOS itself as you say. Is there a reason neither before nor after your PR does the installer use the modules used by NixOS proper to make the menu entries?

Overall I think this looks good, but I don't really feel comfortable merging myself because I *never* work on this part of the repo. I'm sure you've gotten similar responses before :(. I suppose if it did use other shared modules I mentioned above, then we could separate from the mechanism (being able to do these different themed bootloaders), from the policy on on *both* user installations and the installer. The mechanism part I could actually review, and then the policy part would be very small.

On the other hand, you've clearly but a lot of work testing and writing this, and so I feel bad asking you to do a lot more per the above. Maybe we can find who worked on this stuff before (@domenkozar? @grahamc?) and get them to review?

Reply:
> Is there a reason neither before nor after your PR does the installer use the modules used by NixOS proper to make the menu entries?

I wondered the same thing while porting over the configuration; maybe the others working on this part before know why it doesn't use the modules.

* * *

> I suppose if it did use other shared modules I mentioned above, then we could separate from the mechanism

> On the other hand, you've clearly but a lot of work testing and writing this, and so I feel bad asking you to do a lot more per the above

Don't feel bad, I clearly see why using NixOS modules to implement this change would be better! I probably should look into making the grub modules support themes properly, so that it's possible to use the same theme once installed, and then look into using it for the installer module. This, though, shouldn't stop this from being tested and merged. Support for this through module would build upon the same foundations I guess.